### PR TITLE
Added code snippets for h2o model flavor APIs

### PR DESCRIPTION
**Pull request by dmatrix**
Added relevant code snippets for the h20 model flavor APIs.

### CODE REVIEW

**Review by tomasatdatabricks — COMMENTED** [review]
Thanks for the work Jules, this is really helpful!

My only concern is that the code snippets are too big to be embedded in docstring for individual functions. I think it makes the code harder to read and help strings too large. 

I would propose we either a) move the code snippets into an example and link to it in the code or b) move it to the module level instead of individual functions. 

What do you think?

**Comment by dmatrix** [comment]
_My only concern is that the code snippets are too big to be embedded in docstring for individual functions. I think it makes the code harder to read and help strings too large._

This particuarl one is a bit big, compared to other mlflow.<model_flavor>.log_model, for the simple reason
that we really don't need to show them the entire example. Those using h2o model are (and will be) familiar in how to create, train, & fit tha appropriate family of GLMs. So we can cut the code here:
>>> import mlflow
    >>> import mlflow.h2o
    >>> import h2o
    >>> from h2o.estimators.glm import H2OGeneralizedLinearEstimator
    >>> h2o.init()
    >>> #Partial example code used from H20 documentation <http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/early_stopping.html>
    >>> # Initialize and train a GLM
    >>> h2o_model = H2OGeneralizedLinearEstimator(family = 'binomial', early_stopping = True)
    >>> h2o_model.train(x = ... , y = ..., training_frame = ..., validation_frame = valid)
    >>> #log parameters
    >>> mlflow.log_param("early_stopping", "True")
    >>> mlflow.log_param("family", "binomial")
    >>> #log the model
    >>> mlflow.h2o.log_model(h20_model, "h2o_models")

_I would propose we either a) move the code snippets into an example and link to it in the code_
A good compromise is shortend the log_model(...) and keep as it's to be consistent with the rest code
snippets.

I love the idea of having a models/examples/*.py that has a short full example to show use of each module flavor API calls. And then we can link to it for further context. These will provide basic foder as tutorial to show how to use model flavor APIs on your laptop, save the model.<flavor>.deploy(...)

 _or b) move it to the module level instead of individual functions._
API calls with code samples make more sense where we document the API signature, as we do 
with Spark APIs. 

So in short, keep the code snippets succint for now because we need something. And second, for each
model flavor, create a full example.py, demonstrating each call with arguments. Furthermore, this can be
tied quite well with the new SDK for experimenting and tracking. 

So I'm in favor of let's shorten the example, in the h20 case, for shorterm. Longterm, let's think about more extended docs that demonstrate its usage, and then we can link each of this example in the API signature docs.

**Comment by tomasatdatabricks** [comment]
Sounds good, thanks!

**Comment by dmatrix** [comment]
Okay, I will shorten the example and commit.

File: mlflow/h2o.py [hunk 1]
```diff
@@ -60,6 +69,21 @@ def log_model(h2o_model, artifact_path, **kwargs):
...
+    >>> import mlflow.h2o
+    >>> import h2o
+    >>> from h2o.estimators.glm import H2OGeneralizedLinearEstimator
+    >>> #Partial example code used from h2o documentation <http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/early_stopping.html>
+    >>> # Initialize and train a GLM
+    >>> h2o.init()
+    >>> h2o_model = H2OGeneralizedLinearEstimator(family = 'binomial', early_stopping = True)
+    >>> h2o_model.train(...)
+    >>> #log parameters
+    >>> mlflow.log_param("early_stopping", "True")
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: I think we can remove the log param calls.

File: mlflow/h2o.py [hunk 2]
```diff
@@ -60,6 +69,21 @@ def log_model(h2o_model, artifact_path, **kwargs):
     :param h2o_model: H2O model to be saved.
     :param artifact_path: Run-relative artifact path.
     :param kwargs: kwargs to pass to ``h2o.save_model`` method.
+
+    >>> import mlflow
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: If we remove the mlflow.log_param calls below we can remove the import mlflow statement.

File: mlflow/h2o.py [hunk 3]
```diff
@@ -97,6 +121,18 @@ def load_pyfunc(path):
     :param path: Local filesystem path to the model saved by :py:func:`mlflow.h2o.save_model`.
     :rtype: Pyfunc format model with function
             ``model.predict(pandas DataFrame) -> pandas DataFrame``.
+
+    >>> import mlflow
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: I think we do not need import mlflow statement.

File: mlflow/h2o.py [hunk 4]
```diff
@@ -97,6 +121,18 @@ def load_pyfunc(path):
...
+    >>> import mlflow.h2o
+    >>> import h2o
+    >>> # set the path to where the h2o model is saved: local or remote, accessible
+    >>> # from code here
+    >>> h2o.init()
+    >>> h2o_model_dir = ...
+    >>> # set the test Pandas DataFrame
+    >>> pandas_df = ...
+    >>> h2o_model = mlflow.h2o.load_pyfunc(h2o_model_dir)
+    >>> predictions = h2o_model.predict(pandas_df)
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: I think this is a bit confusing in the code.  

The intended way of loading a pyfunc model is to call mlflow.pyfunc.load_pyfunc() which will read the MLmodel yaml file and call the appropriate method to load the model, in this case mlflow.h2o.load_pyfunc(). The important difference is that mlflow.pyfunc.load_pyfunc expects a path to a top level directory containing MLmodel file while the mlflow.h2o.load_pyfunc expects a path to flavor specific data folder. So I think the above code snippet is incorrect.

I don't think we need and example for this function. Instead, we should clarify that it is intended to be used internally.

File: mlflow/h2o.py [hunk 5]
```diff
@@ -109,6 +145,16 @@ def load_model(path, run_id=None):
     :param path: Local filesystem path or run-relative artifact path to the model saved
                  by :py:func:`mlflow.h2o.save_model`.
     :param run_id: Run ID. If provided, combined with ``path`` to identify the model.
+
+    >>> import mlflow
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: I think we do not need the import mlflow statement.

File: mlflow/h2o.py [hunk 6]
```diff
@@ -25,6 +25,15 @@ def save_model(h2o_model, path, conda_env=None, mlflow_model=Model(), settings=N
     :param h2o_model: H2O model to be saved.
     :param path: Local path where the model is to be saved.
     :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+
+    >>> import mlflow
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: I don't think we need the import mlflow statement.

**Review by tomasatdatabricks — COMMENTED** [review]
Thanks for the update, I left few comments.

File: mlflow/h2o.py (hunk 1 above)
Comment (dmatrix) [reply mlflow/h2o.py]: I suppose we could remove it. Will do,

File: mlflow/h2o.py (hunk 2 above)
Comment (dmatrix) [reply mlflow/h2o.py]: Agreed. Will do

File: mlflow/h2o.py (hunk 4 above)
Comment (dmatrix) [reply mlflow/h2o.py]: Now, I'm confused because the API signature for load_pyfunc(path) expects a "path" where model was saved from 'save_model(...) call. Will stop by later today.

File: mlflow/h2o.py (hunk 6 above)
Comment (dmatrix) [reply mlflow/h2o.py]: Yeah, will remove it; redundant.

File: mlflow/h2o.py [hunk 7]
```diff
@@ -60,6 +69,21 @@ def log_model(h2o_model, artifact_path, **kwargs):
     :param h2o_model: H2O model to be saved.
     :param artifact_path: Run-relative artifact path.
     :param kwargs: kwargs to pass to ``h2o.save_model`` method.
+
+    >>> import mlflow
+    >>> import mlflow.h2o
+    >>> import h2o
+    >>> from h2o.estimators.glm import H2OGeneralizedLinearEstimator
+    >>> #Partial example code used from h2o documentation <http://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/early_stopping.html>
```
Comment (tomasatdatabricks) [inline mlflow/h2o.py]: nit: line is too long.

**Review by tomasatdatabricks — COMMENTED** [review]
I think we should remove the code snippet for mlflow.h2o.load_pyfunc as it is currently incorrect and it is being made private by a parallel PR so it might not be worth fixing it.

Other than that, there is just one nit, I left inlined comment.

File: mlflow/h2o.py (hunk 7 above)
Comment (dmatrix) [reply mlflow/h2o.py]: The code snippet for load_pyfunc() has been removed in the last commit? Do you still see it.
I'll break the long line into two lines and commit.

**Comment by mateiz** [comment]
@tomasatdatabricks is this okay to merge now? I restarted the build for it.

**Comment by juntai-zheng** [comment]
Closed as the changes are no longer relevant.
